File: hdc-py/examples/comprehensive.py
"""
完整的使用示例 - 展示所有主要功能
"""

from hdc_rs import HdcClient
import os
import time


def print_section(title):
    """打印章节标题"""
    print("\n" + "=" * 60)
    print(f"  {title}")
    print("=" * 60)


def main():
    print_section("HDC-RS Python 完整示例")
    
    # 1. 连接和设备发现
    print_section("1. 连接到 HDC 服务器")
    try:
        client = HdcClient("127.0.0.1:8710")
        print("✓ 成功连接到 HDC 服务器")
    except Exception as e:
        print(f"✗ 连接失败: {e}")
        return
    
    # 2. 列出设备
    print_section("2. 列出所有设备")
    try:
        devices = client.list_targets()
        print(f"找到 {len(devices)} 个设备:")
        for i, device in enumerate(devices, 1):
            print(f"  {i}. {device}")
    except Exception as e:
        print(f"✗ 列出设备失败: {e}")
        return
    
    if not devices:
        print("\n未找到设备。请确保:")
        print("  1. 设备已连接到计算机")
        print("  2. HDC 服务器正在运行")
        print("  3. 设备已启用 USB 调试")
        return
    
    # 3. 连接到设备
    print_section("3. 连接到设备")
    device_id = devices[0]
    try:
        client.connect_device(device_id)
        print(f"✓ 成功连接到设备: {device_id}")
    except Exception as e:
        print(f"✗ 连接设备失败: {e}")
        return
    
    # 4. 执行 Shell 命令
    print_section("4. 执行 Shell 命令")
    commands = [
        ("获取当前目录", "pwd"),
        ("列出文件", "ls -l /data/local/tmp"),
        ("系统信息", "uname -a"),
        ("当前时间", "date"),
    ]
    
    for description, cmd in commands:
        print(f"\n{description}: {cmd}")
        try:
            output = client.shell(cmd)
            print(output.strip())
        except Exception as e:
            print(f"✗ 错误: {e}")
    
    # 5. 文件传输
    print_section("5. 文件传输")
    
    # 创建测试文件
    test_file = "test_comprehensive.txt"
    with open(test_file, "w") as f:
        f.write(f"测试文件\n")
        f.write(f"创建时间: {time.strftime('%Y-%m-%d %H:%M:%S')}\n")
        f.write(f"设备 ID: {device_id}\n")
    
    print(f"创建测试文件: {test_file}")
    
    # 发送文件
    print("\n发送文件到设备...")
    remote_path = "/data/local/tmp/test_comprehensive.txt"
    try:
        result = client.file_send(test_file, remote_path)
        print(f"✓ {result}")
    except Exception as e:
        print(f"✗ 发送失败: {e}")
    
    # 接收文件
    print("\n从设备接收文件...")
    received_file = "received_comprehensive.txt"
    try:
        result = client.file_recv(remote_path, received_file)
        print(f"✓ {result}")
        
        # 验证内容
        if os.path.exists(received_file):
            with open(received_file, "r") as f:
                print("\n接收到的文件内容:")
                print(f.read())
    except Exception as e:
        print(f"✗ 接收失败: {e}")
    
    # 6. 端口转发
    print_section("6. 端口转发")
    
    print("\n设置端口转发 (local:8888 -> device:8888)...")
    try:
        result = client.fport("tcp:8888", "tcp:8888")
        print(f"✓ {result}")
        
        # 移除端口转发
        print("\n移除端口转发...")
        result = client.fport_remove("tcp:8888 tcp:8888")
        print(f"✓ {result}")
    except Exception as e:
        print(f"✗ 端口转发操作失败: {e}")
    
    # 7. 获取日志
    print_section("7. 获取设备日志")
    
    print("\n获取设备日志 (前 500 个字符)...")
    try:
        logs = client.hilog()
        if logs:
            print(logs[:500])
            print("...")
        else:
            print("日志为空")
    except Exception as e:
        print(f"✗ 获取日志失败: {e}")
    
    # 8. 清理
    print_section("8. 清理")
    
    try:
        # 清理设备上的测试文件
        print("清理设备上的测试文件...")
        client.shell(f"rm -f {remote_path}")
        print("✓ 已清理设备文件")
        
        # 清理本地文件
        if os.path.exists(test_file):
            os.remove(test_file)
        if os.path.exists(received_file):
            os.remove(received_file)
        print("✓ 已清理本地文件")
    except Exception as e:
        print(f"清理时出错: {e}")
    
    print_section("测试完成")
    print("所有主要功能测试完成！")


if __name__ == "__main__":
    main()

File: hdc-rs/examples/comprehensive.rs
//! Comprehensive example demonstrating all HDC features
//!
//! This example shows how to use shell commands, port forwarding, and app management.

use hdc_rs::{ForwardNode, HdcClient};

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    // Initialize logging
    tracing_subscriber::fmt()
        .with_max_level(tracing::Level::INFO)
        .init();

    println!("=== HDC Rust Client - Comprehensive Example ===\n");

    // Step 1: Connect to HDC server
    println!("Step 1: Connecting to HDC server...");
    let mut client = HdcClient::connect("127.0.0.1:8710").await?;
    println!("✓ Connected\n");

    // Step 2: List and select device
    println!("Step 2: Finding devices...");
    let devices = client.list_targets().await?;
    if devices.is_empty() {
        println!("✗ No devices found. Please connect a device and try again.");
        return Ok(());
    }
    println!("✓ Found {} device(s):", devices.len());
    for (i, device) in devices.iter().enumerate() {
        println!("  {}. {}", i + 1, device);
    }

    let device_id = &devices[0];
    println!("\nStep 3: Selecting device: {}", device_id);
    client.connect_device(device_id).await?;
    println!("✓ Device selected\n");

    // Step 4: Execute shell commands
    println!("Step 4: Executing shell commands...");

    println!("  - Getting device info:");
    match client.shell("getprop ro.product.model").await {
        Ok(output) => println!("    Model: {}", output.trim()),
        Err(e) => println!("    Failed: {}", e),
    }

    println!("  - Checking storage:");
    match client.shell("df -h /data").await {
        Ok(output) => {
            for line in output.lines().take(2) {
                println!("    {}", line);
            }
        }
        Err(e) => println!("    Failed: {}", e),
    }

    // Step 5: Port forwarding
    println!("\nStep 5: Managing port forwards...");

    println!("  - Creating TCP forward (8080 -> 8081):");
    match client
        .fport(ForwardNode::Tcp(8080), ForwardNode::Tcp(8081))
        .await
    {
        Ok(response) => println!("    ✓ {}", response.trim()),
        Err(e) => println!("    ✗ {}", e),
    }

    println!("  - Listing all forwards:");
    match client.fport_list().await {
        Ok(tasks) => {
            if tasks.is_empty() {
                println!("    (none)");
            } else {
                for task in &tasks {
                    println!("    - {}", task);
                }
            }
        }
        Err(e) => println!("    ✗ {}", e),
    }

    println!("  - Removing forward:");
    match client.fport_remove("tcp:8080 tcp:8081").await {
        Ok(response) => println!("    ✓ {}", response.trim()),
        Err(e) => println!("    ✗ {}", e),
    }

    // Step 6: App management demo (informational)
    println!("\nStep 6: App management (demo commands)...");
    println!("  Note: These are example commands. Replace paths with actual files.\n");

    println!("  Install app:");
    println!("    client.install(&[\"app.hap\"], InstallOptions::new().replace(true)).await?;");

    println!("\n  Install shared bundle:");
    println!("    client.install(&[\"shared.hsp\"], InstallOptions::new().shared(true)).await?;");

    println!("\n  Uninstall app:");
    println!("    client.uninstall(\"com.example.app\", UninstallOptions::new()).await?;");

    println!("\n  Uninstall but keep data:");
    println!("    client.uninstall(\"com.example.app\", UninstallOptions::new().keep_data(true)).await?;");

    // Step 7: Device logs (hilog)
    println!("\nStep 7: Reading device logs...");
    println!("  - Getting recent logs (with 2 second timeout):");
    match client.hilog(Some("-x")).await {
        Ok(logs) => {
            let lines: Vec<&str> = logs.lines().take(5).collect();
            if lines.is_empty() {
                println!("    (no logs)");
            } else {
                for line in lines {
                    println!("    {}", line);
                }
                let total_lines = logs.lines().count();
                if total_lines > 5 {
                    println!("    ... ({} more lines)", total_lines - 5);
                }
            }
        }
        Err(e) => println!("    Note: {}", e),
    }

    // Summary
    println!("\n=== Summary ===");
    println!("✓ All operations completed successfully!");
    println!("\nFor more examples, see:");
    println!("  - examples/simple_shell.rs - Shell commands");
    println!("  - examples/forward_demo.rs - Port forwarding");
    println!("  - examples/app_demo.rs - App management");
    println!("  - examples/hilog_demo.rs - Device logs");

    Ok(())
}
